src/OrleansRuntime/Startup/StartupBuilder.cs

```diff
@@ -30,30 +44,32 @@ internal static IServiceProvider ConfigureStartup(string startupTypeName, out bo
                 servicesMethod = FindConfigureServicesDelegate(startupType);
                 if (servicesMethod != null && !servicesMethod.MethodInfo.IsStatic)
                 {
-                    usingCustomServiceProvider = true;
+                    hasValidServiceBuilderMethod = true;
                 }
             }
-
-            RegisterSystemTypes(serviceCollection);
-
-            if (usingCustomServiceProvider)
+            Func<IServiceCollection, IServiceProvider> serviceProviderBuilder = null;
+            if (hasValidServiceBuilderMethod)
             {
                 var instance = Activator.CreateInstance(startupType);
-                return servicesMethod.Build(instance, serviceCollection);
+                serviceProviderBuilder = serviceCollection => servicesMethod.Build(instance, serviceCollection);
             }
 
-            return serviceCollection.BuildServiceProvider();
+            return ConfigureStartup(serviceProviderBuilder, configureServices, out usingCustomServiceProvider);
         }
 
-        private static void RegisterSystemTypes(IServiceCollection serviceCollection)
+        private static IServiceProvider ConfigureStartup(Func<IServiceCollection, IServiceProvider> serviceProviderBuilder, Action<IServiceCollection> configureServices, out bool usingCustomServiceProvider)
         {
-            // Register the system classes and grains in this method.
-            // Note: this method will probably have to be moved out into the Silo class to include internal runtime types.
+            IServiceCollection serviceCollection = new ServiceCollection();
+            configureServices(serviceCollection);
 
-            serviceCollection.AddTransient<GrainBasedMembershipTable>();
-            serviceCollection.AddTransient<GrainBasedReminderTable>();
-        }
+            usingCustomServiceProvider = serviceProviderBuilder != null;
+            if (usingCustomServiceProvider)
+            {
+                return serviceProviderBuilder(serviceCollection);
+            }
```

Comment:
The user could create their own container (from their library of choice), override the default Orleans ones, and return that. Given the method signature (IServiceCollection -> IServiceProvider) I don't see another way yet. I'm not sure if I should be configuring those default providers in a different way to make that easier.